packages/cli/src/lib/snapshot/snapshot.ts

```diff
@@ -138,45 +139,45 @@ export class Snapshot {
     });
   }
 
-  public async waitUntilDone(
-    operationToWaitFor?: ResourceSnapshotsReportType | null,
-    options: WaitUntilDoneOptions = {},
-    iteratee = (_report: ResourceSnapshotsReportModel) => {}
+  public waitUntilDone(
+    operationToWaitFor: ResourceSnapshotsReportType | null,
+    options: Partial<WaitUntilDoneOptions> = {},
+    onRetryCb = (_report: ResourceSnapshotsReportModel) => {}
   ) {
+    const opts = {...Snapshot.defaultWaitOptions, options};
     const toMilliseconds = (seconds: number) => seconds * 1e3;
-    const opts = {
-      ...Snapshot.defaultWaitOptions,
-      ...options,
-    };
-    let timeout: NodeJS.Timeout;
-
-    await new Promise<void>((resolve, reject) => {
-      if (opts.wait > 0) {
-        timeout = setTimeout(() => {
-          const err = new SnapshotOperationTimeoutError(this);
-          clearInterval(interval);
-          reject(err);
-        }, toMilliseconds(opts.wait));
+
+    return retry(
+      this.waitUntilDoneRetryFunction(operationToWaitFor, onRetryCb),
+      {
+        retries: Infinity,
+        minTimeout: toMilliseconds(opts.waitInterval),
+        maxTimeout: toMilliseconds(opts.waitInterval),
+        maxRetryTime: options.wait,
       }
+    );
+  }
 
-      const interval = setInterval(async () => {
-        await this.refreshSnapshotData();
+  private waitUntilDoneRetryFunction(
+    operationToWaitFor: ResourceSnapshotsReportType | null,
+    onRetryCb: (report: ResourceSnapshotsReportModel) => void
+  ): () => Promise<void> {
+    return (async () => {
+      await this.refreshSnapshotData();
 
-        const isExpectedOperation =
-          !operationToWaitFor || this.latestReport.type === operationToWaitFor;
+      const isExpectedOperation =
+        !operationToWaitFor || this.latestReport.type === operationToWaitFor;
 
-        const isOngoing = Snapshot.ongoingReportStatuses.includes(
-          this.latestReport.status
-        );
+      const isOngoing = Snapshot.ongoingReportStatuses.includes(
+        this.latestReport.status
+      );
 
-        iteratee(this.latestReport);
+      onRetryCb(this.latestReport);
 
-        if (!isOngoing && isExpectedOperation) {
-          clearTimeout(timeout);
-          clearInterval(interval);
-          resolve();
-        }
-      }, toMilliseconds(opts.waitInterval));
-    });
+      if (isOngoing || !isExpectedOperation) {
```

Comment:
Reversed the condition using De Morgan Law to keep it simple. No return needed in the end